### DISCUSSION TITLE
Setting Up a Dedicated RU Machine for srsRAN with USRP B210

### DISCUSSION THREAD

**Discussion by justinlau21** [srsran/srsRAN_Project#959] [General Help]
I have successfully set up a 5G testbed with the following configuration:

Machine 1 (192.168.0.109): Installed with Open5GS (handles core network functions).
Machine 2 (192.168.0.116): Installed with srsRAN (vanilla installation), and the UHD driver is also installed for USRP communication.
Machine 3 (192.168.0.120): Installed with UHD driver and connected to a USRP B210 via USB 3.0.
So far, everything works perfectly when the USRP B210 is plugged directly into Machine 2. However, I want to use Machine 3 as a dedicated Radio Unit (RU) machine that communicates with Machine 2 (where srsRAN DU is running) via the network.

The Issue:
I need to configure Machine 3 (connected to the USRP B210) to run the RU functionality. Currently, I am unable to get the RU running on Machine 3, and it seems that Machine 2 (where the DU and CU are running) is not properly connecting to the USRP B210 on Machine 3.

I have already tried the following steps:

I modified the srsRAN gNB configuration (gnb.yml) to point to Machine 3 (IP address 192.168.0.120) as the RU machine.
The configuration file for srsRAN (gnb.yml) includes the correct addr and device_args for the USRP B210 (e.g., addr=192.168.0.120, type=b200).
I ensured that Machine 2 and Machine 3 can communicate over the network (verified via ping).

This suggests that Machine 2 is unable to communicate with the USRP B210 on Machine 3. I am unsure if the problem is with the network configuration, device addressing, or the need for some additional configuration on Machine 3 to expose the USRP device remotely.

What I’ve Tried:
Network Configuration:

Ensured that both Machine 2 and Machine 3 are on the same subnet and can communicate (verified with ping).
The USRP B210 is directly connected to Machine 3 via USB 3.0.
UHD and srsRAN Configuration:

On Machine 3, I ran uhd_find_devices, which successfully detects the USRP B210.
On Machine 2, I have tried setting the device_args in the gnb.yml file to use addr=192.168.0.120, but I still get the same error.
UHD Network Support:

I’m not sure if I need to configure UHD for network access (or if it’s already enabled by default). If UHD needs any extra configuration (e.g., enabling network mode), I would appreciate any guidance.

What I Need Help With:
How to configure Machine 3 as a dedicated RU machine to expose the USRP B210 to Machine 2 over the network.
Correct configuration of gnb.yml to point to the correct IP address and ensure srsRAN on Machine 2 can connect to the USRP B210 on Machine 3.
Any additional network or software configuration needed on Machine 3 to allow the USRP B210 to be accessed remotely by Machine 2.

**Comment by pgawlowicz** [comment]
You do not need Machine 3 in your setup; the B210 already acts as RU and has its IP address.

**Reply by justinlau21** [reply]
Hi, i'm working on a 5G fronthaul testbed for research purpose where i want to create a setup like 7.2 split, but i don't have access to the O-RU needed in the setup.

**Reply by pgawlowicz** [reply]
We do have split 7.2 support using in-house OFH library. However, it can be used only with commercial o-RUs. Currently, there is no way to connect USRP via OFH.

**Reply by justinlau21** [reply]
If i use a emulated RU, can i connect UE (physical UE or srsUE, etc...) to it?

**Reply by pgawlowicz** [reply]
I do not think so. 
@sauka Could you comment?

**Reply by sauka** [reply]
hey @justinlau21 ,
No, the `ru_emulator` doesn't allow to connect a real UE. It is meant for evaluating performance of the srsRAN CU/DU in a 7.2 split scenario using the **test mode**. For example, you could run the srsRAN gNB in the Machine 2 using test mode and ru_emulator in Machine 3 and observe the achieved throughput, monitor the logs etc.

Regarding your original question, I am not sure it is possible to use B210 as a network device, AFAIK it only supports USB3.0 and won't be visible to Machine 2. To confirm this you can run `uhd_find_devices` **in Machine 2** with the args you tried in gnb.yml and see if it detects your B210

**Reply by justinlau21** [reply]
Can a USRP N210 works?

USRP N210 connected to the network switch.

**Reply by pblanz3** [reply]
Hi! I'm having the same problem @justinlau21. I don't know if a USRP N210 can works, because I "just" have USRP B210.
Were you able to fix the problem?
